hazelcast/src/main/java/com/hazelcast/config/MerkleTreeConfig.java

```diff
@@ -0,0 +1,258 @@
+/*
+ * Copyright (c) 2008-2018, Hazelcast, Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.hazelcast.config;
+
+import com.hazelcast.nio.ObjectDataInput;
+import com.hazelcast.nio.ObjectDataOutput;
+import com.hazelcast.nio.serialization.IdentifiedDataSerializable;
+import com.hazelcast.spi.annotation.Beta;
+import com.hazelcast.util.StringUtil;
+
+import java.io.IOException;
+
+import static com.hazelcast.util.Preconditions.checkNotNull;
+
+/**
+ * Configuration for a merkle tree.
+ * The merkle tree is a data structure used for efficient comparison of the
+ * difference in the contents of large data structures. The precision of
+ * such a comparison mechanism is defined by the depth of the merkle tree.
+ * <p>
+ * A larger depth means that a data synchronization mechanism will be able
+ * to pinpoint a smaller subset of the data structure contents in which a
+ * change occurred. This causes the synchronization mechanism to be more
+ * efficient. On the other hand, a larger tree depth means the merkle tree
+ * will consume more memory.
```

Comment:
is there an option to put merkle tree inside HD?

Reply:
Most of the data (the entry keys) can be kept on off-heap as `Data` in `com.hazelcast.util.collection.OAHashSet#table`. The on-heap part is mostly `int` hash-codes which are all of the other nodes - `com.hazelcast.wan.merkletree.AbstractMerkleTreeView#tree`. So the on-heap memory usage is not too big.